server/src/main/java/org/elasticsearch/cluster/coordination/CoordinationDiagnosticsService.java

```diff
@@ -588,6 +622,223 @@ private Scheduler.Cancellable fetchClusterFormationInfo(
+     * @param masterEligibleNode The node being polled
+     * @param responseConsumer The response consumer to be wrapped
+     * @param cancellableConsumer The list of Cancellables
+     * @param isCancelled If true, this task has been cancelled in another thread and does not need to do any more work
+     * @return A wrapped Consumer that will run fetchCoordinationDiagnostics()
+     */
+    private Consumer<RemoteMasterHealthResult> rescheduleDiagnosticsFetchConsumer(
+        DiscoveryNode masterEligibleNode,
+        Consumer<RemoteMasterHealthResult> responseConsumer,
+        Consumer<Scheduler.Cancellable> cancellableConsumer,
+        AtomicBoolean isCancelled
+    ) {
+        return response -> {
+            if (isCancelled.get()) {
+                logger.trace("An attempt to reschedule a cancelled remote coordination diagnostics task is being ignored");
+            } else {
+                /*
+                 * We make sure that the task hasn't been cancelled. If it has, we don't reschedule the task. Since this block is not
+                 * synchronized it is possible that the task will be cancelled after the check and before the following is run. In that case
+                 * the cancellableConsumer will immediately cancel the task.
+                 */
+                cancellableConsumer.accept(
+                    fetchCoordinationDiagnostics(
+                        masterEligibleNode,
+                        responseConsumer.andThen(
+                            rescheduleDiagnosticsFetchConsumer(masterEligibleNode, responseConsumer, cancellableConsumer, isCancelled)
+                        ),
+                        isCancelled
+                    )
+                );
+            }
+        };
+    }
+
+    /**
+     * This method returns quickly, but in the background schedules to query the remote node's cluster diagnostics in 10 seconds
+     * unless cancel() is called on the Cancellable that this method returns.
+     * @param node The node to poll for cluster diagnostics
+     * @param responseConsumer The consumer of the cluster diagnostics for the node, or the exception encountered while contacting it
+     * @param isCancelled If true, this task has been cancelled in another thread and does not need to do any more work
+     * @return A Cancellable for the task that is scheduled to fetch cluster diagnostics
+     */
+    private Scheduler.Cancellable fetchCoordinationDiagnostics(
+        DiscoveryNode node,
+        Consumer<RemoteMasterHealthResult> responseConsumer,
+        AtomicBoolean isCancelled
+    ) {
+        StepListener<Releasable> connectionListener = new StepListener<>();
+        StepListener<CoordinationDiagnosticsAction.Response> fetchCoordinationDiagnosticsListener = new StepListener<>();
+        long startTime = System.nanoTime();
+        connectionListener.whenComplete(releasable -> {
+            if (isCancelled.get()) {
+                IOUtils.close(releasable);
```

Comment:
It could. But previously we were using `Releasables.close()`, which also can throw an exception (it catches the IOException and rethrows it as an `UncheckedIOException` so that it doesn't cause compilation errors with the streams API but the result is the same. In this block I think it won't matter because this poll was cancelled anyway, so we don't need to make sure to reschedule anything. In the block below where `Releasables.close()` is used, I believe the exception will be caught by `fetchCoordinationDiagnosticsListener`'s exception handler, and the task will be rescheduled.